### PR TITLE
GEODE-8741_1: implementation of API changes

### PR DESCRIPTION
**Pull request by mivanac**
Added proposal of implementation, of new API for block/suppress CQ Create/Update/Destroy notifications.
Idea is that user can chose which notifications wants to suppress, with goal to  remove unnecessary notifications, and improve performanse.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [*] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [*] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [*] Is your initial contribution a single, squashed commit?

- [*] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/cache/query/CqQuery.java
Comment (kohlmu-pivotal) [inline geode-core/src/main/java/org/apache/geode/cache/query/CqQuery.java:164]: @mivanac I quite like the idea of possibly limiting what notifications one would like to receive. I think that the approach is maybe a little limiting, as it is specifically focussed on  "update" only. Could we achieve a much more encompassing solution for delete or create as well? Maybe an inclusions or exclusions list of events that the end result would like to receive?

Given that this is an external facing API, I would encourage changes to be more encompassing, in order to avoid many API changes/deprecations of external facing APIs.
@upthewaterspout @agingade @nabarunnag could anyone of you possibly be able to work with @mivanac to come up with an API that is possibly a little more future proofed.

**Review by kohlmu-pivotal — CHANGES_REQUESTED** [review]

**Review by DonalEvans — COMMENTED** [review]
Rather than using a bitmask and requiring users to specify a non-intuitive integer value to indicate which notifications are suppressed, might it be better to create a wrapper class with setter and getter methods for each notification type which can be passed to the `newCq()` method? This adds a little complexity in implementation, but would be far more user-friendly.

**Review by DonalEvans — COMMENTED** [review]
Instead of storing the notification suppression internally as an integer, would it be simpler to just use the new `CqSuppressNotification` class everywhere, and just check the value of the boolean flags rather than doing bit arithmetic? If there's a specific reason for preferring the current implementation that's fine though.

**Comment by mivanac** [comment]
Main motivation is to simplify implementation, and also simplify adding and getting part to/from message ExecuteCQ.

File: geode-core/src/main/java/org/apache/geode/cache/query/CqQuery.java
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/cache/query/CqQuery.java:162]: I think this is supposed to be Geode 1.14?

**Review by kirklund — COMMENTED** [review]
Please make sure you're creating new unit tests (not just distributed/integration tests) for classes you touch that have not unit test. At a minimum, I'd recommend covering the method(s) you add or alter.

**Comment by onichols-pivotal** [comment]
This PR has been open for some while with no recent activity.  Please keep Geode tidy by closing PRs you're not actively using.

**Comment by mivanac** [comment]
This PR is waiting on review from CODEOWNERS.

**Comment by mivanac** [comment]
Hi  @jdeppe-pivotal, @jinmeiliao, @agingade, @DonalEvans, @nabarunnag, @jchen21, and @kirklund, could you review this PR.
Thanks.

**Review by DonalEvans — COMMENTED** [review]
After doing a little research, I think that the simplest and clearest implementation for the "which notification to suppress" information might be an `EnumSet`. This would prevent the need for doing bit masking and passing around an int, would remove the need for individual "isXSuppressed()" methods (since you can just call `set.contails()`), and would be more intuitive from a development point of view. Here's an example of what I'm thinking of:
```
class Scratch {
  enum SuppressedNotifications {
    CREATE, UPDATE, DESTROY;
  }
  public static void main(String[] args) {
    EnumSet<SuppressedNotifications> onlyAllowDestroy = EnumSet.of(SuppressedNotifications.CREATE, SuppressedNotifications.UPDATE);
    for (SuppressedNotifications type : SuppressedNotifications.values()) {
      System.out.println("Set contains type " + type + " = " + onlyAllowDestroy.contains(type));
    }
  }
}
```
which outputs:
```
Set contains type CREATE = true
Set contains type UPDATE = true
Set contains type DESTROY = false
```
It's possible that this implementation may be slightly more performant, as an `EnumSet` is backed by a bit array which is the size of the number of enum values it's based on (3 in this case) instead of being a 32-bit int. I apologise for not suggesting this initially, but I only just took the time to really do some digging and see if there was a better implementation for this case.

**Comment by mivanac** [comment]
Thanks for the comments and proposal. I will apply it.

**Comment by agingade** [comment]
I am having hard time understanding the requirement here...The jira ticket just mention block CQ notification, what kind of notification we are talking here...By design CQ does not raise any notification...Can you be more specific about the notification you are mentioning about...
Also, i don't think it should be managed by the create/destroy APIs or by adding entirely new set of APIs, thats redundant.

**Review by agingade — CHANGES_REQUESTED** [review]
I have added my comment about the requirement of this as I am unable to find details in the Jira ticket or here.

**Comment by agingade** [comment]
I have added following comments in the ticket also:
I have to look into the PR to see what your are really are trying to achieve in this ticket. It will be helpful if you can add more information with the example (depending on the issue).

E.g. 
Avoid sending CQ events for specific CQ query-operations (E.g. destroy Query operation) to client.
For a region.put() which results in CQ event with its query operation as destroy; avoid sending this event to client.

I am not seeing any value/scenario that needs this kind of behavior. Do you have any example can you please explain here.

The core CQ design aspect was to establish a data view representing a query result that can be expressed in the form of query.  And support event mechanism that will enable to manage/maintain that view as the change happens on the source data. The changes requested in this ticket, breaks the core design and can allow one to represent/establish a different data view that is not in-line with the source data (a query result view). I don't see need for supporting that kind of behavior in the server side. If Client is not interested to see specific event/operation it should be filtering this out on the client side. 

Why the need for this behavior?
Looking the Pull Request opened; its not like its avoiding processing of CQs, which is generally more expensive part of having CQ; only thing its avoiding sending the CQ event to client. If this is the behavior client/application is looking for; probably CQ is not the functionality to use.

Also, it will be nice to use Request For Change (RFC) process for any issues that has behavior/functional or new API requirements. It will help to get feedback during design phase itself.

**Comment by agingade** [comment]
From an implementation point of view; this seems to be added/handled in CqQuery as a CqAttributes...
E.g:
interestedEventsWithQueryOperations = { ALL, CREATE, UPDATE, DESTROY}

**Comment by mivanac** [comment]
Thanks for comments. I agree with you that there is no gain from this PR. I will close these PRs.

**Pull request closed without merging**
